src/lib.rs

```diff
@@ -212,56 +242,62 @@ impl<O> IO<O> {
 /// assert_eq!(result, "meow\n");
 /// # Ok (()) }
 /// ```
-pub fn pure <T> (value: T) -> IO<T> {
+pub fn pure <'a, T> (value: T) -> IO<'a, T> where T: 'a {
     IO (Pure (value))
 }
 
 /// Ignore the output value of `x`.
-pub fn void <T> (x: IO<T>) -> IO<()>
+pub fn void <'a, T> (x: IO<'a, T>) -> IO<'a>
 where
     T: 'static
 {
     x.replace(())
 }
 
 /// Print a line to stdout.
-pub fn println (s: impl ToString) -> IO<()> {
+pub fn println (s: impl ToString) -> IO<'static> {
     struct PrintLn (String);
 
-    impl Action for PrintLn {
+    impl Action<'static> for PrintLn {
         type Output = ();
         fn perform (&mut self, Context { o, .. }: Context<'_>) -> Result<Self::Output> {
             writeln!(o, "{}", self.0)
         }
     }
-    
+
     IO::new(PrintLn (s.to_string()))
 }
 
 /// Read a file's contents into a string.
-pub fn read_file (path: impl AsRef<Path> + 'static) -> IO<String> {
+pub fn read_file <'a> (path: impl AsRef<Path> + 'a) -> IO<'a, String> {
     let mut p = Some (path);
-    IO::new(move || {
+    IO::from_fn(move |_| {
         let p = p.take().expect("Re-evaluation of IO action!");
         let s = fs::read_to_string(p)?;
-        Ok (pure (s))
+        Ok (s)
     })
 }
 
 /// Acquire a resource and apply a function to it.
-pub fn bracket <I, O, F> (
-    acquire: IO<I>,
-    release: IO<()>,
-) -> impl FnOnce (F) -> IO<O>
+pub fn bracket <'a, I, O, F> (
```

Comment:
I'm struggling to understand what this function does, could you explain it in layman's terms?

Reply:
@MaeIsBad I have updated the documentation, feedback would be appreciated!

Reply:
Looks great 👍